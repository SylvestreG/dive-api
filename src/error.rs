use thiserror::Error;

#[derive(Error, Debug)]
pub enum DiveErr {
    #[error("diesel Error")]
    IoErr(#[from] diesel::result::Error),
    #[error("r2d2 Error")]
    R2d2Err(#[from] r2d2::Error),
    #[error("dotenv Error")]
    DotenvError(#[from] dotenvy::Error),
    #[error("uuid Erro")]
    UuidError(#[from] uuid::Error),
}
