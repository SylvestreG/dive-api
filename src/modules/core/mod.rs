pub mod fixtures;
pub mod security;
