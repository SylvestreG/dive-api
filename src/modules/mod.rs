pub mod core;
pub mod qualification;
pub mod users;
