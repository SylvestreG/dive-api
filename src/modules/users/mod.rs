mod fixtures;
mod models;
mod repositories;
pub mod services;
#[cfg(test)]
mod tests;
