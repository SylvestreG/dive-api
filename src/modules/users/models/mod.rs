pub mod level;
pub mod user;
pub mod user_level;
