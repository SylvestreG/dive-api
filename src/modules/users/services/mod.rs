pub mod level;
pub mod user;
