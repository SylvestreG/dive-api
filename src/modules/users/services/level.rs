use crate::error::DiveErr;
use crate::modules::users::{fixtures, models, repositories};
use diesel::r2d2::ConnectionManager;
use diesel::PgConnection;
use r2d2::Pool;
use uuid::Uuid;

pub struct Service {
    repository: repositories::level::Repository,
}

impl Service {
    #[must_use]
    pub fn new(con: Pool<ConnectionManager<PgConnection>>) -> Self {
        Service {
            repository: repositories::level::Repository::new(con),
        }
    }

    #[must_use]
    pub fn fixtures(&self) -> Box<fixtures::level::Fixtures> {
        Box::new(fixtures::level::Fixtures::new(self.repository.clone()))
    }

    pub fn find_by_uuid(&self, uuid: Uuid) -> Result<models::level::Entity, DiveErr> {
        // TODO ACL

        self.repository.get(uuid)
    }

    pub fn list_all_levels(&self) -> Result<Vec<models::level::Entity>, DiveErr> {
        // TODO ACL

        self.repository.all()
    }

    #[must_use]
    pub fn user_by_role(&self) -> Vec<models::level::Entity> {
        vec![]
    }
}
