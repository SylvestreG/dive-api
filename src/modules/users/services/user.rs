use crate::modules::users::{fixtures, models, repositories};
use diesel::r2d2::ConnectionManager;
use diesel::PgConnection;
use r2d2::Pool;
use uuid::Uuid;
use crate::error::DiveErr;

pub struct Service {
    repository: repositories::user::Repository,
}

impl Service {

    #[must_use]
    pub fn new(con: Pool<ConnectionManager<PgConnection>>) -> Self {
        Service {
            repository: repositories::user::Repository::new(con),
        }
    }

    #[must_use]
    pub fn fixtures(&self) -> Box<fixtures::user::Fixtures> {
        Box::new(fixtures::user::Fixtures::new(self.repository.clone()))
    }

    #[must_use]
    pub fn find_by_uuid(&self, _uuid: Uuid) -> models::user::Entity {
        todo!()
    }

    #[must_use]
    pub fn list_all_user(&self) -> Result<Vec<models::user::Entity>, DiveErr> {
        self.repository.all()
    }

    #[must_use]
    pub fn user_by_role(&self) -> Vec<models::user::Entity> {
        vec![]
    }
}
