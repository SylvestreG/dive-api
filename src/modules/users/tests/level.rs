use crate::error::DiveErr;
use crate::modules::users;
use crate::modules::users::services;
use diesel::r2d2::ConnectionManager;
use diesel::PgConnection;
use dotenvy::dotenv;
use r2d2::Pool;
use test_context::{test_context, TestContext};

struct LevelTests {
    svc: services::level::Service,
}

impl TestContext for LevelTests {
    fn setup() -> LevelTests {
        dotenv().unwrap();

        let db_url = std::env::var("DATABASE_URL").unwrap();
        let manager = ConnectionManager::<PgConnection>::new(db_url);
        let pool = Pool::builder().build(manager).unwrap();

        LevelTests {
            svc: services::level::Service::new(pool),
        }
    }

    fn teardown(self) {
        // Perform any teardown you wish.
    }
}

#[test_context(LevelTests)]
#[test]
fn list_all_level(ctx: &mut LevelTests) -> Result<(), DiveErr> {
    let mut levels = ctx.svc.list_all_levels()?;

    levels.sort();
    insta::assert_json_snapshot!(levels, {
        "[].created_at" => "[date]",
        "[].updated_at" => "[date]",
    });
    Ok(())
}

#[test_context(LevelTests)]
#[test]
fn get_level(ctx: &mut LevelTests) -> Result<(), DiveErr> {
    let lvl_1 = ctx
        .svc
        .find_by_uuid(users::fixtures::user::DATA[1].id.unwrap())?;
    let lvl_2 = ctx
        .svc
        .find_by_uuid(users::fixtures::user::DATA[1].id.unwrap())?;

    insta::assert_json_snapshot!(lvl_1, {
        ".created_at" => "[date]",
        ".updated_at" => "[date]",
    });
    insta::assert_json_snapshot!(lvl_2, {
        ".created_at" => "[date]",
        ".updated_at" => "[date]",
    });
    Ok(())
}
