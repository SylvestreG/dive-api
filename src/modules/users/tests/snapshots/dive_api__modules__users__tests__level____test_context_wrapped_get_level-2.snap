---
source: src/modules/users/tests/level.rs
expression: lvl_2
---
{
  "id": "f633eb82-ff98-4e93-9264-a27c1bb54c87",
  "level_name": "Niveau 2",
  "created_at": "[date]",
  "updated_at": "[date]"
}
